tools/resourcedocsgen/cmd/docs/registry.go

```diff
@@ -126,31 +129,71 @@ func genResourceDocsForAllRegistryPackages(registryRepoPath, baseDocsOutDir, bas
 		return errors.Wrap(err, "reading the registry packages dir")
 	}
 
-	for _, f := range metadataFiles {
-		glog.Infof("=== starting %s ===\n", f.Name())
-		glog.Infoln("Processing metadata file")
-		metadataFilePath := filepath.Join(registryPackagesPath, f.Name())
+	var m sync.Mutex
+	var errs []error
+	addError := func(err error) {
+		m.Lock()
+		defer m.Unlock()
+		errs = append(errs, err)
+	}
 
-		b, err := os.ReadFile(metadataFilePath)
-		if err != nil {
-			return errors.Wrapf(err, "reading the metadata file %s", metadataFilePath)
-		}
+	// Concurrency is limited via a checkout-checkin system.
+	//
+	// Each time a process starts, it checks out a hallpass from the monitor (blocking
+	// if none are available). When the process completes, it returns it's hallpass to
+	// the monitor.
+	//
+	// Because there are a fixed number of hallpasses (maxConcurrency), and each
+	// process needs a hallpass to run, we limit effective concurrency to the number
+	// of available hallpasses.
+	maxConcurrency := runtime.NumCPU()
+
+	var hallpass struct{}
+	monitor := make(chan struct{}, maxConcurrency)
+	for i := 0; i < maxConcurrency; i++ {
+		monitor <- hallpass
+	}
 
-		var metadata pkg.PackageMeta
-		if err := yaml.Unmarshal(b, &metadata); err != nil {
-			return errors.Wrapf(err, "unmarshalling the metadata file %s", metadataFilePath)
-		}
+	var w sync.WaitGroup
+	w.Add(len(metadataFiles))
+	for _, f := range metadataFiles {
+		go func(f os.DirEntry) {
+			defer func(hallpass struct{}) {
+				w.Done()
+				monitor <- hallpass
+			}(<-monitor)
+			glog.Infof("=== starting %s ===\n", f.Name())
+			glog.Infoln("Processing metadata file")
+			metadataFilePath := filepath.Join(registryPackagesPath, f.Name())
+
+			b, err := os.ReadFile(metadataFilePath)
+			if err != nil {
+				addError(errors.Wrapf(err, "reading the metadata file %s", metadataFilePath))
+				return
+			}
 
-		docsOutDir := filepath.Join(baseDocsOutDir, metadata.Name, "api-docs")
-		err = genResourceDocsForPackageFromRegistryMetadata(metadata, docsOutDir, basePackageTreeJSONOutDir)
-		if err != nil {
-			return errors.Wrapf(err, "generating resource docs using metadata file info %s", f.Name())
-		}
+			var metadata pkg.PackageMeta
+			if err := yaml.Unmarshal(b, &metadata); err != nil {
+				addError(errors.Wrapf(err, "unmarshalling the metadata file %s", metadataFilePath))
+				return
+			}
+
+			docsOutDir := filepath.Join(baseDocsOutDir, metadata.Name, "api-docs")
+			err = genResourceDocsForPackageFromRegistryMetadata(metadata, docsOutDir, basePackageTreeJSONOutDir)
+			if err != nil {
+				addError(errors.Wrapf(err, "generating resource docs using metadata file info %s", f.Name()))
+				return
+			}
 
-		glog.Infof("=== completed %s ===", f.Name())
+			glog.Infof("=== completed %s ===", f.Name())
```

Comment:
(This is my first time seeing this custom codegen logic so take this with a grain of salt.)

AFAICT we always generate docs for each metadata file we visit? If that's true we should try to short-circuit if the metadata hasn't changed and outputs already exist, e.g. by tracking a hash or mtime somewhere.

If we turn `genResourceDocsForPackageFromRegistryMetadata` into a self-contained binary we could probably handle all of this with `make -j`.

Reply:
> We need to since we don't store the output in repo.

Let's check it in then! No need to re-compute docs that haven't changed.

> If/when we implement caching, this will still generate wins when the cache key changes (say when we bump the pulumi/pulumi version).

For sure, the regenerate-everything case will always exist. I'm thinking more about the local dev loop where incremental builds are important.